// @shivrajgodle/mydemo-lib/esm2015/lib/counter/counter.component.js
import { Component, EventEmitter, Output } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/material/button";
import * as i2 from "@angular/material/icon";
export class CounterComponent {
    constructor() {
        this.countChanged = new EventEmitter();
        this.clickCount = 0;
    }
    ngOnInit() {
    }
    onPlusClick() {
        this.clickCount++;
        this.countChanged.emit(this.clickCount);
    }
    onMinusClick() {
        this.clickCount--;
        this.countChanged.emit(this.clickCount);
    }
}
CounterComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.5", ngImport: i0, type: CounterComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
CounterComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.0.5", type: CounterComponent, selector: "lib-counter", outputs: { countChanged: "countChanged" }, ngImport: i0, template: "\n\n\n  <button mat-mini-fab color=\"accent\" aria-label=\"Example icon button with a plus one icon\" (click)=\"onPlusClick()\">\n    <mat-icon>add</mat-icon>\n  </button>\n  \n  <input type=\"text\" readonly class=\"counter\" value=\"{{clickCount}}\">\n\n  <button mat-mini-fab color=\"accent\" aria-label=\"Example icon button with a plus one icon\" (click)=\"onMinusClick()\">\n    <mat-icon>remove</mat-icon>\n  </button>\n", styles: [".counter{text-align:center;background-color:#fff;width:50px;height:33.5px;border-top:1px solid #000;border-bottom:1px solid #000}"], components: [{ type: i1.MatButton, selector: "button[mat-button], button[mat-raised-button], button[mat-icon-button],             button[mat-fab], button[mat-mini-fab], button[mat-stroked-button],             button[mat-flat-button]", inputs: ["disabled", "disableRipple", "color"], exportAs: ["matButton"] }, { type: i2.MatIcon, selector: "mat-icon", inputs: ["color", "inline", "svgIcon", "fontSet", "fontIcon"], exportAs: ["matIcon"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.5", ngImport: i0, type: CounterComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'lib-counter',
                    templateUrl: './counter.component.html',
                    styleUrls: ['./counter.component.css']
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { countChanged: [{
                type: Output
            }] } });
