// auth-guard/dist/lib/util/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = exports.LogColor = void 0;
var tslib_1 = require("tslib");
var LogColor;
(function (LogColor) {
    LogColor["Reset"] = "\u001B[0m";
    LogColor["Bright"] = "\u001B[1m";
    LogColor["Dim"] = "\u001B[2m";
    LogColor["Underscore"] = "\u001B[4m";
    LogColor["Blink"] = "\u001B[5m";
    LogColor["Reverse"] = "\u001B[7m";
    LogColor["Hidden"] = "\u001B[8m";
    LogColor["FgBlack"] = "\u001B[30m";
    LogColor["FgRed"] = "\u001B[31m";
    LogColor["FgGreen"] = "\u001B[32m";
    LogColor["FgYellow"] = "\u001B[33m";
    LogColor["FgBlue"] = "\u001B[34m";
    LogColor["FgMagenta"] = "\u001B[35m";
    LogColor["FgCyan"] = "\u001B[36m";
    LogColor["FgGray"] = "\u001B[37m";
    LogColor["FgWhite"] = "\u001B[97m";
    LogColor["BgBlack"] = "\u001B[40m";
    LogColor["BgRed"] = "\u001B[41m";
    LogColor["BgGreen"] = "\u001B[42m";
    LogColor["BgYellow"] = "\u001B[43m";
    LogColor["BgBlue"] = "\u001B[44m";
    LogColor["BgMagenta"] = "\u001B[45m";
    LogColor["BgCyan"] = "\u001B[46m";
    LogColor["BgWhite"] = "\u001B[47m";
})(LogColor = exports.LogColor || (exports.LogColor = {}));
var Logger = /** @class */ (function () {
    function Logger() {
    }
    Logger.getTimeString = function () {
        var date = new Date();
        return ("[" + this.formatDateTimeString(date.getDate()) + "." + //
            (this.formatDateTimeString(date.getMonth() + 1) + ".") + //
            (date.getFullYear() + " -- ") + //
            (this.formatDateTimeString(date.getHours()) + ":") + //
            (this.formatDateTimeString(date.getMinutes()) + ":") + //
            (this.formatDateTimeString(date.getSeconds()) + "]"));
    };
    Logger.formatDateTimeString = function (toFormat) {
        return ("0" + toFormat).slice(-2);
    };
    Logger.info = function (color) {
        var message = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            message[_i - 1] = arguments[_i];
        }
        console.log.apply(console, tslib_1.__spreadArray(tslib_1.__spreadArray(["" + color + this.getTimeString() + ":"], message), ['\x1b[0m']));
    };
    Logger.log = function () {
        var message = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            message[_i] = arguments[_i];
        }
        this.info.apply(this, tslib_1.__spreadArray([LogColor.FgWhite], message));
    };
    Logger.warn = function () {
        var message = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            message[_i] = arguments[_i];
        }
        this.info.apply(this, tslib_1.__spreadArray(["" + LogColor.FgYellow + LogColor.Bright], message));
    };
    Logger.debug = function () {
        var message = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            message[_i] = arguments[_i];
        }
        this.info.apply(this, tslib_1.__spreadArray([LogColor.FgGreen], message));
    };
    Logger.error = function () {
        var message = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            message[_i] = arguments[_i];
        }
        this.info.apply(this, tslib_1.__spreadArray(["" + LogColor.FgRed + LogColor.Bright], message));
    };
    return Logger;
}());
exports.Logger = Logger;
