// @shivrajgodle/mydemo-lib/esm2015/lib/buttion/buttion.component.js
import { Component, Input } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/material/button";
import * as i2 from "@angular/material/icon";
export class ButtionComponent {
    // @Output() buttonClickEmit: EventEmitter<string> = new EventEmitter<string>();
    constructor() { }
    ngOnInit() {
    }
    onButtonClick() {
        // this.buttonClickEmit.emit('you have clicked on button');
    }
}
ButtionComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.5", ngImport: i0, type: ButtionComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
ButtionComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "12.0.5", type: ButtionComponent, selector: "lib-buttion", inputs: { buttonText: "buttonText", iconName: "iconName" }, ngImport: i0, template: "\n\n\n\n<button type=\"button\" (click)=\"onButtonClick()\" mat-raised-button color=\"primary\">\n\n  {{buttonText}}\n\n  <mat-icon>{{iconName}} </mat-icon>\n\n</button>\n\n\n", styles: [""], components: [{ type: i1.MatButton, selector: "button[mat-button], button[mat-raised-button], button[mat-icon-button],             button[mat-fab], button[mat-mini-fab], button[mat-stroked-button],             button[mat-flat-button]", inputs: ["disabled", "disableRipple", "color"], exportAs: ["matButton"] }, { type: i2.MatIcon, selector: "mat-icon", inputs: ["color", "inline", "svgIcon", "fontSet", "fontIcon"], exportAs: ["matIcon"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.5", ngImport: i0, type: ButtionComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'lib-buttion',
                    templateUrl: './buttion.component.html',
                    styleUrls: ['./buttion.component.css']
                }]
        }], ctorParameters: function () { return []; }, propDecorators: { buttonText: [{
                type: Input
            }], iconName: [{
                type: Input
            }] } });
