// @shivrajgodle/mydemo-lib/esm2015/lib/mydemo-lib.module.js
import { NgModule } from '@angular/core';
import { MydemoLibComponent } from './mydemo-lib.component';
import { CounterComponent } from './counter/counter.component';
import { MatButtonModule } from '@angular/material/button';
import { MatIconModule } from '@angular/material/icon';
import { ButtionComponent } from './buttion/buttion.component';
import { CCInputBoxComponent } from './ccinput-box/ccinput-box.component';
import { MatFormFieldModule } from '@angular/material/form-field';
import { ReactiveFormsModule } from '@angular/forms';
import { MatInputModule } from '@angular/material/input';
import * as i0 from "@angular/core";
export class MydemoLibModule {
}
MydemoLibModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.0.5", ngImport: i0, type: MydemoLibModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MydemoLibModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.0.5", ngImport: i0, type: MydemoLibModule, declarations: [MydemoLibComponent,
        CounterComponent,
        ButtionComponent,
        CCInputBoxComponent], imports: [MatButtonModule,
        MatIconModule,
        MatFormFieldModule,
        ReactiveFormsModule,
        MatInputModule], exports: [MydemoLibComponent, CounterComponent, ButtionComponent, CCInputBoxComponent] });
MydemoLibModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.0.5", ngImport: i0, type: MydemoLibModule, imports: [[
            MatButtonModule,
            MatIconModule,
            MatFormFieldModule,
            ReactiveFormsModule,
            MatInputModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.0.5", ngImport: i0, type: MydemoLibModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        MydemoLibComponent,
                        CounterComponent,
                        ButtionComponent,
                        CCInputBoxComponent
                    ],
                    imports: [
                        MatButtonModule,
                        MatIconModule,
                        MatFormFieldModule,
                        ReactiveFormsModule,
                        MatInputModule
                    ],
                    exports: [
                        MydemoLibComponent, CounterComponent, ButtionComponent, CCInputBoxComponent
                    ]
                }]
        }] });
